### PR TITLE
[RISCV] Add 16 bit GPR sub-register for Zhinx.

### PR DESCRIPTION
**Pull request by topperc**
This patches adds a 16 bit register class for use with Zhinx instructions. This makes them more similar to Zfh instructions and allows us to only spill 16 bits.

I've added CodeGenOnly instructions for load/store using GPRF16 as that gave better results than insert_subreg/extract_subreg. I'm using FSGNJ for GPRF16 copy with Zhinx as that gave better results. Zhinxmin will use ADDI+subreg operations.

Function arguments use this new GPRF16 register class for f16 arguments with Zhinxmin. Eliminating the need to use RISCVISD::FMV* nodes.

I plan to extend this idea to Zfinx next.

After that, I want to try to extend this to 32 bit integer W instructions. My thought is that we can arrange to have all writes to the 32 bit GPR guarantee sign extension similar to how Mips64 is handled. Unfortunately, we are missing some W instructions in Zba and Zbs that would make this straightforward. I want to see if this is more viable than my previous attempt at making i32 a legal type which kept it inside the existing 64-bit GPR class.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVRegisterInfo.td [hunk 1]
```diff
@@ -83,41 +83,81 @@ def sub_gpr_odd  : SubRegIndex<32, 32> {
...
+  def X25_H : RISCVReg<25,"x25", ["s9"]>;
+  def X26_H : RISCVReg<26,"x26", ["s10"]>;
+  def X27_H : RISCVReg<27,"x27", ["s11"]>;
+  def X28_H : RISCVReg<28,"x28", ["t3"]>;
+  def X29_H : RISCVReg<29,"x29", ["t4"]>;
+  def X30_H : RISCVReg<30,"x30", ["t5"]>;
+  def X31_H : RISCVReg<31,"x31", ["t6"]>;
+  }
+
+  let SubRegIndices = [sub_16] in {
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVRegisterInfo.td:123]: Do we need `CoveredBySubRegs`?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVRegisterInfo.td:123]: It's not covered by subregs. There's no sub reg for the upper bits.

**Comment by preames** [comment]
This goes beyond what I can meaningfully review, but I'll note that I see nothing to particularly object to here.  

Naming wise, _H is, I assume, for half.  Since we'll probably use these for bf16, and possible someday i16, maybe pick a different name suffix?

**Comment by jrtc27** [comment]
FLH and LH both use half as the term, and the format of the float shouldn’t affect the load or store, just the interpretation for arithmetic purposes, so _H seems like the appropriate thing to use for me?

**Comment by preames** [comment]
> FLH and LH both use half as the term, and the format of the float shouldn’t affect the load or store, just the interpretation for arithmetic purposes, so _H seems like the appropriate thing to use for me?

Good point, naming suggestion withdrawn.

**Review by asb — APPROVED** [review]
LGTM in that the reasoning makes sense to me and I can't see problems from going through the code. That said, it's been a while since I added a new register class.

I think it would be nice to have a brief comment in RISCVRegisterInfo.td explaining why we're defining this register class and how it's used. Perhaps that makes more sense once you make similar changes for zfinx, but in general I find such in-tree documentation of intent really helpful.

You say that fsgnj gives better results vs mv - I'm assuming this is for real world code? Because for the modified in-tree tests, surely using `mv` where possible is better due to being potentially compressible? Happy to go with your judgement, but it probably deserves a TODO/FIXME somewhere like the tests to note that a semantically equivalent mv would be more compressible (and so someone sufficiently motivated might later pick it up).

EDIT to add:
> After that, I want to try to extend this to 32 bit integer W instructions. My thought is that we can arrange to have all writes to the 32 bit GPR guarantee sign extension similar to how Mips64 is handled. Unfortunately, we are missing some W instructions in Zba and Zbs that would make this straightforward. I want to see if this is more viable than my previous attempt at making i32 a legal type which kept it inside the existing 64-bit GPR class.

This definitely sounds like an interesting direction. And if we can demonstrate adding those missing instructions makes life easier for the compiler, it's probably worth at least discussing in e.g. the scalar efficiency working group.

**Comment by topperc** [comment]
> You say that fsgnj gives better results vs mv - I'm assuming this is for real world code? Because for the modified in-tree tests, surely using `mv` where possible is better due to being potentially compressible? Happy to go with your judgement, but it probably deserves a TODO/FIXME somewhere like the tests to note that a semantically equivalent mv would be more compressible (and so someone sufficiently motivated might later pick it up).

The subreg operations needed to use the existing ADDI seem to pessimize Machine Copy Propagation on some of the lit tests. I think this is because some implicit operands get added and MachineCopyPropagation doesn't know how those implicit operands relate to the explicit operands so it conservatively gives up. I have not done any testing of this on real world code. I suspect a GPRF16 CodeGenOnly copy instruction may fix this issue, but haven't checked yet.

**Review by asb — APPROVED** [review]

**Comment by topperc** [comment]
@asb can you re-review this? I added compressed versions of the load/store CodeGenOnly instructions, a pseudo for mv, updated MakeCompressible, MergeBaseOffset.

File: llvm/lib/Target/RISCV/RISCVCallingConv.cpp [hunk 2]
```diff
@@ -139,6 +139,23 @@ ArrayRef<MCPhysReg> RISCV::getArgGPRs(const RISCVABI::ABI ABI) {
   return ArrayRef(ArgIGPRs);
 }
 
+static ArrayRef<MCPhysReg> getArgGPR16s(const RISCVABI::ABI ABI) {
+  // The GPRs used for passing arguments in the ILP32* and LP64* ABIs, except
+  // the ILP32E ABI.
+  static const MCPhysReg ArgIGPRs[] = {RISCV::X10_H, RISCV::X11_H, RISCV::X12_H,
+                                       RISCV::X13_H, RISCV::X14_H, RISCV::X15_H,
+                                       RISCV::X16_H, RISCV::X17_H};
+  // The GPRs used for passing arguments in the ILP32E/ILP64E ABI.
```
Comment (asb) [inline llvm/lib/Target/RISCV/RISCVCallingConv.cpp:148]: Typo: ILP64E => LP64E (I've just committed a fix to other similar instances in this file).

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 3]
```diff
@@ -1505,6 +1520,9 @@ unsigned RISCVInstrInfo::getInstSizeInBytes(const MachineInstr &MI) const {
   }
 
   switch (Opcode) {
+  case RISCV::PseudoMV_FPR16INX:
+    // MV is always compressible.
```
Comment (asb) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:1524]: This comment made me double check, so sharing that process: Although we'd expect `c.li` to be used instead, `c.mv` can't handle an X0 operand. But from what I can see, we expand to ADDI and we have the obvious CompressPat from ADDI with X0 operand to c.li.

Perhaps expand the comment to "MV is always compressible (to either c.mv or c.li rd, 0)."

**Review by asb — APPROVED** [review]
LGTM! Sorry for missing your request to re-review. I'm still missing the setup I had with Phabricator Herald :( But I'll try to improve my workflow.

**Pull request merged**